### ISSUE TITLE
[flang][hlfir] fir.char_convert verification failure

### ISSUE THREAD

**Issue by vzakhari** [llvm/llvm-project#64315]
Reproducer:
```
subroutine test(c,n)
  character(*,4),intent(in) :: c(:)
  integer,intent(in) :: n
  interface
     subroutine callee(c)
       character(*),intent(in) :: c(:)
     end subroutine callee
  end interface
  call show([character(n)::c])
end subroutine test

subroutine test2(x)
  integer,intent(in) :: x
  character(kind=4) :: cx
  cx = achar(x)
end subroutine test2
```

flang-new -flang-experimental-hlfir repro.f90:
```
error: loc("repro.f90":9:3): 'fir.char_convert' op operand #0 must be any reference, but got '!fir.boxchar<4>'
error: loc("repro.f90":15:3): 'fir.char_convert' op operand #0 must be any reference, but got '!hlfir.expr<!fir.char<1>>'
```

**Comment by vzakhari** [comment]
Hi @cabreraam, can you please take a look?

**Comment by cabreraam** [comment]
Fixed in 561fbe441ea919b97301dede408ccdf2a0a29b49
